chainerx_cc/chainerx/CMakeLists.txt

```diff
@@ -115,68 +115,52 @@ if(MSVC)
     add_library(chainerx STATIC ${chainerx_srcs})
 else()
     add_library(chainerx_base STATIC ${chainerx_srcs})
-    add_library(chainerx SHARED empty.cc)
+    add_library(chainerx STATIC empty.cc)
 endif()
 
 # abseil
 target_link_libraries(
-    chainerx
+    chainerx_base
     PUBLIC
     absl::failure_signal_handler
     absl::bad_optional_access
     PRIVATE
     absl::flat_hash_map
     )
 
-if(MSVC)
-    target_link_libraries(chainerx
-        PUBLIC
-        ${CMAKE_DL_LIBS})
+# dlopen / dlclose
+target_link_libraries(chainerx PUBLIC ${CMAKE_DL_LIBS})
 
-    target_link_options(chainerx PUBLIC /wholearchive:$<TARGET_FILE:chainerx_routines>)
-    target_link_options(chainerx PUBLIC /wholearchive:$<TARGET_FILE:chainerx_native>)
-    target_link_options(chainerx PUBLIC /wholearchive:$<TARGET_FILE:chainerx_testing>)
-    target_link_libraries(chainerx PRIVATE chainerx_routines)
-    target_link_libraries(chainerx PRIVATE chainerx_native)
-    target_link_libraries(chainerx PRIVATE chainerx_testing)
+# ChainerX
+set(chainerx_sub_libs
+    chainerx_base
+    chainerx_routines
+    chainerx_native
+    chainerx_testing
+    )
 
-    if(${CUDA_FOUND})
-      target_link_options(chainerx PUBLIC /wholearchive:$<TARGET_FILE:chainerx_cuda>)
-      target_link_libraries(chainerx PRIVATE chainerx_cuda)
-    endif()
+if(${CUDA_FOUND})
+    set(chainerx_sub_libs ${chainerx_sub_libs} chainerx_cuda)
+endif()
 
+if(MSVC)
+    target_link_libraries(chainerx PRIVATE ${chainerx_sub_libs})
+    foreach(lib ${chainerx_sub_libs})
+        target_link_options(chainerx PUBLIC /wholearchive:$<TARGET_FILE:${lib}>)
+    endforeach(lib)
 elseif(${APPLE})
-    target_link_libraries(chainerx
-        PUBLIC
-        ${CMAKE_DL_LIBS})
-
-    target_link_libraries(chainerx PRIVATE -Wl,-force_load chainerx_base)
-    target_link_libraries(chainerx PRIVATE -Wl,-force_load chainerx_routines)
-    target_link_libraries(chainerx PRIVATE -Wl,-force_load chainerx_native)
-    target_link_libraries(chainerx PRIVATE -Wl,-force_load chainerx_testing)
-
-    if(${CUDA_FOUND})
-      target_link_libraries(chainerx PRIVATE -Wl,-force_load chainerx_cuda)
-    endif()
-
+    foreach(lib ${chainerx_sub_libs})
+        target_link_libraries(chainerx PRIVATE -Wl,-force_load ${lib})
+    endforeach(lib)
     target_link_libraries(chainerx PRIVATE -Wl,-noall_load)
 else()
     target_link_libraries(chainerx
-        PUBLIC
-        ${CMAKE_DL_LIBS}
         PRIVATE
-        -Wl,--whole-archive)
-
-    target_link_libraries(chainerx PRIVATE chainerx_base)
-    target_link_libraries(chainerx PRIVATE chainerx_routines)
-    target_link_libraries(chainerx PRIVATE chainerx_native)
-    target_link_libraries(chainerx PRIVATE chainerx_testing)
-
-    if(${CUDA_FOUND})
-        target_link_libraries(chainerx PRIVATE chainerx_cuda)
-    endif()
-
-    target_link_libraries(chainerx PRIVATE -Wl,--no-whole-archive)
+        -Wl,--allow-multiple-definition
```

Comment:
Better to comment why this is necessary? You usually should not need this?

Reply:
Thank you for a comment.
Yes, I think it shouldn't be necessary, but without it MNIST example fails to build for duplicate abseil symbols.
Maybe I need some more investigation.

```/usr/bin/c++   -g -Wall -Wextra -fPIC -pthread -fopenmp -O0   -pthread -rdynamic CMakeFiles/train_mnist.dir/mnist.o CMakeFiles/train_mnist.dir/train_mnist.o  -o train_mnist -Wl,-rpath,/home/niboshi/.cudnn/active/cuda/lib64:/usr/local/cuda/lib64 ../../chainerx/libchainerx.a -Wl,--whole-archive ../../chainerx/libchainerx_base.a ../../abseil-build/absl/debugging/libabsl_absl_failure_signal_handler.a ../../abseil-build/absl/debugging/libabsl_absl_examine_stack.a ../../abseil-build/absl/hash/libabsl_absl_hash.a ../../abseil-build/absl/types/libabsl_absl_bad_variant_access.a ../../abseil-build/absl/hash/libabsl_absl_city.a ../../abseil-build/absl/container/libabsl_absl_raw_hash_set.a ../../abseil-build/absl/types/libabsl_absl_bad_optional_access.a ../../abseil-build/absl/container/libabsl_absl_hashtablez_sampler.a ../../abseil-build/absl/synchronization/libabsl_absl_synchronization.a ../../abseil-build/absl/debugging/libabsl_absl_stacktrace.a ../../abseil-build/absl/debugging/libabsl_absl_symbolize.a ../../abseil-build/absl/debugging/libabsl_absl_debugging_internal.a ../../abseil-build/absl/debugging/libabsl_absl_demangle_internal.a ../../abseil-build/absl/synchronization/libabsl_absl_graphcycles_internal.a ../../abseil-build/absl/base/libabsl_absl_malloc_internal.a ../../abseil-build/absl/time/libabsl_absl_time.a ../../abseil-build/absl/strings/libabsl_absl_strings.a ../../abseil-build/absl/strings/libabsl_absl_strings_internal.a ../../abseil-build/absl/numeric/libabsl_absl_int128.a ../../abseil-build/absl/time/libabsl_absl_civil_time.a ../../abseil-build/absl/time/libabsl_absl_time_zone.a ../../abseil-build/absl/base/libabsl_absl_throw_delegate.a ../../abseil-build/absl/base/libabsl_absl_base.a ../../abseil-build/absl/base/libabsl_absl_dynamic_annotations.a ../../abseil-build/absl/base/libabsl_absl_log_severity.a ../../abseil-build/absl/base/libabsl_absl_spinlock_wait.a ../../chainerx/routines/libchainerx_routines.a ../../chainerx/native/libchainerx_native.a ../../chainerx/testing/libchainerx_testing.a ../../googletest-build/googlemock/gtest/libgtestd.a ../../chainerx/cuda/libchainerx_cuda.a -pthread /usr/local/cuda/lib64/libcudart_static.a -lrt -lcublas /usr/local/cuda/lib64/libcudart_static.a -lrt -lcublas -ldl /home/niboshi/.cudnn/active/cuda/lib64/libcudnn.so /usr/local/cuda/lib64/libcusolver.so -Wl,--no-whole-archive 
../../abseil-build/absl/base/libabsl_absl_log_severity.a(log_severity.o): In function `absl::lts_2019_08_08::NormalizeLogSeverity(absl::lts_2019_08_08::LogSeverity)':
/home/niboshi/w/repos/chainer/chainerx_cc/build/abseil/absl/base/log_severity.cc:22: multiple definition of `absl::lts_2019_08_08::operator<<(std::ostream&, absl::lts_2019_08_08::LogSeverity)'
../../abseil-build/absl/base/libabsl_absl_base.a(log_severity.o):/home/niboshi/w/repos/chainer/chainerx_cc/build/abseil/absl/base/log_severity.cc:22: first defined here
```

Reply:
It seems it's a problem of abseil.
Using the master branch of abseil, this error is gone.
But now I'm experiencing errors from linking duplicate `libcudart_static.a`.

Reply:
I'm not able to fix the CUDA library issue.
I undid the upgrade of abseil and wrote a comment.

Reply:
In my env, libcudart_static.a issue didn't reproduced: https://gist.github.com/take-cheeze/8cdad2310b23e32072fc4c519ccb2a39
How about trying `set(CUDA_USE_STATIC_CUDA_RUNTIME OFF)` ?

Reply:
https://github.com/abseil/abseil-cpp/commit/d9aa92d7fb324314f9df487ac23d32a25650b742
https://github.com/abseil/abseil-cpp/commit/fa8c75182fbfdeddb2485fc0d53baeda3f40b7a3
Seems to fix the log_severity issue

Reply:
Even if CUDA issue can be solved, it's not necessarily good to upgrade abseil to some untagged commit.
Isn't `--allow-multiple-definition` better?

Reply:
`--allow-multiple-definition` may cause some trouble that is difficult to investigate so I don't want to use it.
The cost of updating abseil afterward seems more low instead of using this flag.

Reply:
Could you elaborate on your concern?

Reply:
I assume the effect of the flag is local to chainerx, and it doesn't affect users that link to it. May be wrong.

Reply:
My concern is that it may hide duplicate symbol error of linker which should be checked.
Though doing this only in `_core.so` should to be fine.

Reply:
I see, that's a problem, of course.😅
`CUDA_USE_STATIC_CUDA_RUNTIME` worked fine.
All of the three fixes in this commit 69e2a7f79187cc42817b00360aa3c857746eabc0  were needed to build `mnist` example